### ISSUE TITLE
perf: Close per-model gaps surfaced by AIsEval fair-comparison vs PyTorch (LSTM, Transformer, MLP, regression-pipeline, OpenCL RSS)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#436]
> ## 📊 Status (cross-repo) — updated 2026-05-26
>
> This is an umbrella spanning **two repos**. The tensors-side primitives are done; the perf *wins* land when the framework layers call them (tracked in the framework repo).
>
> | Item | P | Repo | Status |
> |---|---|---|---|
> | OpenCL kernel-cache RSS on CPU-only (GPU-only init) | P2 | Tensors | ✅ done — #437 |
> | CNN inference lead — perf gate (preserve) | — | Tensors | ✅ done — #437 |
> | Fused `LstmSequenceForward` primitive | P0 | Tensors | ✅ done — #437 (8.19 ms < PyTorch 11.76 ms) |
> | Fused `MultiHeadAttentionForward` primitive | P0 | Tensors | ✅ done — #437 (8.5 ms < PyTorch 13.85 ms) |
> | Fused `MlpForward` primitive | P1 | Tensors | ✅ primitive done — #437; ⏳ **perf gate + verify** in progress |
> | Wire `LSTMLayer`/`MHALayer`/`DenseLayer.Forward` → primitives | P0/P1 | **ooples/AiDotNet** | ⏳ tracked in ooples/AiDotNet#1447 |
> | `AiModelBuilder<MultipleRegression>` 48k overhead | P2 | **ooples/AiDotNet** | ⏳ tracked in ooples/AiDotNet#1447 |
> | Fused **backward** for LSTM/MHA (training path) | P1 | Tensors | ❌ forward-only today; follow-up |
> | Gradient-step (backward+optimizer) perf | P1 | both | ❌ gated on #433 |
>
> **Why #437 didn't auto-close this:** #436 is a multi-repo umbrella; #437 intentionally closed only the tensors-side primitives (see its "What this PR does not close" section). The framework-side wiring is the remaining blocker for the AIsEval rerun to show the gaps closed.
>
> ---
>
## Context

External repo [lyudmilpetrov/AIsEval](https://github.com/lyudmilpetrov/AIsEval) was publishing a PyTorch-vs-AiDotNet benchmark whose numbers were not apples-to-apples (full \`AiModelBuilder\` lifecycle vs a 5-line \`torch.linalg.lstsq\` wrapper; hand-rolled MLP for every model name; \`GC.GetTotalMemory()\` vs \`psutil\` RSS). We landed [lyudmilpetrov/AIsEval#25](https://github.com/lyudmilpetrov/AIsEval/pull/25) to make both sides symmetric and ran the rebuilt benchmark on a CPU-only Windows 11 rig. The fair-comparison numbers now surface real gaps we need to close before the next time someone re-runs it.

Raw report data: \`Reporting/aidotnet-mct.json\`, \`Reporting/pytorch.json\`, \`Reporting/pytorch-lstm-reduced.json\`, \`Reporting/regression-bench.json\` in PR #25.

Goal: be **<= 1.0x PyTorch** on every model family the AIsEval scaffold exercises, on CPU, at PyTorch's default workload (\`--epochs 3 --train-batches 20 --batch-size 64 --inference-iterations 100 --warmup-iterations 10\`). One model already wins; the rest need work.

### Inference @ bs=128 (steady-state latency)

| Model | PyTorch | AiDotNet | Ratio | Status |
|---|---|---|---|---|
| MLP | 1.91 ms | 8.94 ms | **4.7x slower** | gap |
| **CNN** | **41.42 ms** | **6.21 ms** | **6.7x faster** | **lead — preserve** |
| LSTM | 11.76 ms | did not finish | **>>100x slower** | severe gap |
| Transformer | 13.85 ms | 233.25 ms | **16.8x slower** | severe gap |

### Training (3e x 20b x bs=64, total wall time)

| Model | PyTorch | AiDotNet | Ratio |
|---|---|---|---|
| MLP | 1.38 s | 1.41 s | ~tied |
| CNN | 1.63 s | 2.87 s | 1.8x slower |
| LSTM | 1.87 s | did not finish | >>50x slower at scaffold defaults |
| Transformer | 5.91 s | 16.18 s | 2.7x slower |

### Per-batch backward time (gradient_seconds_avg)

| Model | PyTorch | AiDotNet | Ratio |
|---|---|---|---|
| MLP | 5.21 ms | 16.78 ms | 3.2x slower |
| CNN | 12.04 ms | 44.39 ms | 3.7x slower |
| Transformer | 20.34 ms | 182.70 ms | 9.0x slower |

### REST regression endpoint (full \`AiModelBuilder<MultipleRegression>\` vs \`nn.Linear+MSELoss+Adam\` 200 epochs)

| Dataset | PyTorch | AiDotNet | Ratio |
|---|---|---|---|
| Small (99 rows x 10 features), warm avg | 515 ms | 505 ms | ~tied |
| Large (47 999 rows x 10 features) | 3.87 s | 16.79 s | **4.3x slower** |

### Memory (peak RSS during training)

| Model | PyTorch | AiDotNet |
|---|---|---|
| MLP | 282 MB | 478 MB |
| CNN | 291 MB | 2 914 MB |
| Transformer | 323 MB | 3 507 MB |

The AiDotNet excess RSS is dominated by the OpenCL backend kernel cache (~591 kernels compiled at startup) even when no GPU is present.

### P0 — severe inference gap, blocks any \"AiDotNet is fast\" claim

- [ ] **LSTM \`Predict()\` forward path** — \`LSTMLayer\` forward over \`[B=128, seq=32, features=32]\` does not complete a single inference iteration in <2 s on this rig; the full inference loop (92 forward passes) was still running after 3+ minutes when I killed it. Training itself is competitive (1.37 s for 1ep x 3b x bs=32 vs PyTorch 1.70 s on the same slice), so the bottleneck is specifically in the recurrent forward kernel called from \`Predict\`, not the autograd path. PyTorch routes \`nn.LSTM\` to OneDNN/MKL under the hood; we have no equivalent kernel.
  - **File:** \`src/AiDotNet.Tensors/.../LSTMLayer.cs\` Forward path (search for \"_hiddenSize\", \"timeSteps\" loop).
  - **Acceptance:** LSTM \`Predict\` at \`[128, 32, 32]\` <= 25 ms (within 2x PyTorch's 11.76 ms).

- [ ] **Transformer \`Predict()\` @ bs=128** — 16.8x slower (233 ms vs 13.85 ms). Multi-head attention forward is the suspect. Cross-references #404 (parallel pack-B in RunParallelUnsafe — wide-N FFN/Attn shapes) — likely contributes but not sufficient alone.
  - **Acceptance:** Transformer \`Predict\` at \`[128, 32, 32]\` <= 25 ms (within 2x PyTorch's 13.85 ms).

### P1 — MLP/Dense dispatch + backward perf

- [ ] **MLP \`Predict()\` @ bs=128** — 4.7x slower (8.94 ms vs 1.91 ms) on \`Dense(784->512) -> Dense(512->128) -> Dense(128->10)\`. With 468k params and bs=128, this is purely 3 GEMMs in (128, 784, 512), (128, 512, 128), (128, 128, 10). At those shapes we should be at PyTorch parity or better. Likely a dispatch-overhead issue rather than peak compute. Cross-references #408 (GEMV bypass), #371 (small-shape pack-free streaming path).
  - **Acceptance:** MLP \`Predict\` at \`[128, 784]\` <= 3 ms (within 1.6x PyTorch's 1.91 ms).

- [ ] **Gradient-step time** — backward+optimizer per batch is 3-9x PyTorch. Worst on Transformer (182.7 ms vs 20.3 ms). Cross-references #433 (tape backward + optimizer 15x forward on 25M-param Paraformer) which targets the same code path; gating this on the same work landing.
  - **Acceptance:** Each \`gradient_seconds_avg\` from the AIsEval scaffold within 2x of PyTorch.

### P2 — high-level lifecycle + memory

- [ ] **\`AiModelBuilder<MultipleRegression>\` at 48k rows** — 4.3x slower than PyTorch's \`nn.Linear+MSELoss+Adam\` 200-epoch loop (16.79 s vs 3.87 s). This is the lifecycle that \`/api/Regression/SimpleRegression\` exercises — \`ConfigureDataLoader + ConfigureModel + BuildAsync\` with internal 70/15/15 split + validation + final predict. Likely contributors: validation phase compute, internal split allocations, builder ceremony overhead per call. Filed under AiDotNet.Tensors because the underlying solve goes through tensor ops; may need cross-filing in \`ooples/AiDotNet\` for builder-level changes.
  - **Acceptance:** Large-set wall time <= 6 s (within 1.6x PyTorch).

- [ ] **OpenCL kernel cache RSS on CPU-only runs** — ~2-3 GB of kernel cache is compiled at process startup even when no GPU is present, inflating measured RSS by 5-10x vs PyTorch. A user who installs AiDotNet on a CPU-only box and runs one small model still pays this. Should be lazy / opt-in / skipped when no compatible GPU is detected.
  - **File:** \`src/AiDotNet.Tensors/Engines/DirectGpu/OpenCL/OpenClBackend.cs\` initialization.
  - **Acceptance:** A CPU-only \`dotnet run -- --models mlp\` should not compile OpenCL kernels and should have RSS <= 600 MB peak for the MLP workload above.

### Preserve

- [ ] **CNN inference @ bs=128** — currently **6.7x faster than PyTorch** (6.21 ms vs 41.42 ms). Add a regression test or perf gate so this doesn't slip when we touch the conv path.

## Reproducer

\`\`\`bash
# AiDotNet 4-model scaffold (skip lstm for now — does not finish at defaults)
cd aidotnet-benchmarks && dotnet run --no-build -c Release -- \
    --models mlp,cnn,transformer --output ../results/aidotnet-mct.json

# PyTorch scaffold
cd ../pytorch-benchmarks && python -m pytorch_benchmarks \
    --models mlp,cnn,lstm,transformer --output ../results/pytorch.json

# REST regression endpoint comparison
cd ../aidotnet-benchmarks
ASPNETCORE_URLS=http://127.0.0.1:7000 dotnet bin/Release/net10.0/AiDotNetBenchmarks.dll &
cd ../pytorch-benchmarks
python -m uvicorn pytorch_benchmarks.api:app --host 127.0.0.1 --port 8000 &
cd .. && python Reporting/run-regression-bench.py
\`\`\`

Raw JSONs from the run that produced the table above are committed to PR #25 under \`Reporting/\`.

## Cross-references

- #368 (BlasManaged mega) and sub-tasks #369-#373, #404, #408 — should pick up the MLP/Transformer GEMM-dispatch gaps as a side effect.
- #433 (tape backward + optimizer perf) — should close the gradient-step gap.
- #358 (ConvTranspose2D L2 below MKL peak) — adjacent conv perf work; unrelated to CNN here (CNN already wins) but in the same family.
- Upstream PR with the rerunable benchmark: https://github.com/lyudmilpetrov/AIsEval/pull/25.

**Comment by ooples** [comment]
## Fresh head-to-head verification (PR #474)

Built a same-machine head-to-head harness (AiDotNet fused primitives vs TorchSharp/MKL) at the exact AIsEval shapes, with a strict **p95(ours) < median(torch)** bar and min-of-7-rounds noise robustness. Numbers on a 16-core dev box:

| Model | AiDN med | AiDN p95 | Torch med | Torch p95 | Verdict |
|---|---:|---:|---:|---:|:-:|
| MLP | 1.37 ms | 1.75 ms | ~0.63 ms | ~1.03 ms | LOSS |
| Transformer (MHA) | 5.31 ms | 7.25 ms | 2.74 ms | 3.87 ms | LOSS |
| LSTM | 5.76 ms | 6.63 ms | 2.78 ms | 3.65 ms | LOSS |

**Big picture:** the fused primitives already closed most of the original catastrophic gaps (MLP 8.94 ms, LSTM "didn't finish", Transformer 233 ms). What remains is **GEMM kernel quality**, attributed precisely:

- **MLP** — zero-alloc raw 3-GEMM floor is ~1.2 ms at optimal threads, already ≥ torch's whole MLP median. `FusedLinear` uses native **OpenBLAS**; torch uses **MKL**. Landed a thread-cap win (median 1.65→1.37, p95 3.24→1.75) but beating MKL outright needs a faster kernel → #475.
- **MHA** — 7.3 MB/call alloc (internal SDPA scores/softmax) + managed `SimdGemm` quality; scales fine with threads → #476.
- **LSTM** — thread-insensitive (flat ~5.8 ms 1..16 cores); unfused 32-step recurrent kernel vs torch's fused sequence LSTM → #477.

PR #474 lands the MLP thread cap + the reproducible diagnostics (`--ab-aiseval-h2h`, `--ab-aiseval-floor`, `--ab-aiseval-diag`, `--ab-aiseval-dopsweep`). Deep-kernel work to actually clear the strict bar is split into #475 (MLP/MKL), #476 (MHA/SDPA), #477 (LSTM fused).

**Comment by ooples** [comment]
Tensors-side **MlpForward perf gate + verify** (the last ⏳ Tensors row in the status table) is done in **PR #541**.

The pre-existing gate was `[Fact(Skip=…)]`, which xUnit always skips even with a filter — so it never actually executed. PR #541 converts it to the repo's runnable perf-gate convention (`Category=Perf` + `AIDOTNET_RUN_PERF_GATES=1`), switches to a per-iter median, and gives it a per-TFM budget. Verified Release:

| TFM | median | budget | result |
|---|---|---|---|
| net10.0 | 0.55 ms | 4 ms | ✅ |
| net471 | 4.84 ms (no-intrinsics scalar floor) | 7 ms | ✅ |

Both well under the 8.94 ms unfused-dispatch path the gate guards. Leaving this umbrella **open** — the remaining work is framework-side wiring (ooples/AiDotNet#1447) + the fused-backward/gradient-step rows, not Tensors primitives.
